### ISSUE TITLE
Project-level .npmrc not respected

### ISSUE THREAD

**Issue by andybee** [architect/architect#1061]
When using tree-shaking to bundle your dependencies in each Lambda function on deployment, if you use a private NPM package within `shared` and have a `.npmrc` file at the project root level, the triggered `npm install` during deploy will not find the private package.

**Expected behavior**
I'd prefer to keep any `.npmrc` at the same level as the related `package.json` which, in this case, is the root project level. I presume what might be happening is an inability to find this because npm is run within the function's path/scope?

**Comment by ryanblock** [comment]
Yeah, what you described is accurate, npm runs from within that directory. We can change this up, but I don't currently have any elegant ideas of how to handle this, especially considering the side effects that may occur if we start doing things with the `.nmprc` file within the project. Perhaps the best path forward here is to just advise that in userland either put that `.nmprc` specifically where it's needed, or configure it globally so those settings are available no matter which dir Hydrate is working in?

**Comment by andybee** [comment]
Appreciate the concern and also can't offer an "easy" solution, but the two workaround both present issues for me:

1. The dependency is in `shared`, so I'd have to put the `.npmrc` in every function folder...?
2. Making it global it becomes something that needs explaining to any future developer that gets onboarded, rather than something that's just handled by checking out the repo, installing and running up Sandbox.

**Comment by ryanblock** [comment]
No, you'd only have to put an npmrc file in shared – Hydrate installs in shared, then copies to each dir iirc.

**Comment by andybee** [comment]
OK, that feels like a reasonable workaround to commit the `.npmrc` in to the shared folder.

Not sure if to keep this issue open or not as the underlying issue could still potentially trip up new users, where the newer tree-shaking methodology encourages everything in one package.json, but you need to be wary of actual usage in a specific function if doing something unusual with `.npmrc`? I figure it's an edge case and the documentation tweak might be the best bet.

**Comment by andybee** [comment]
Just to add I can't see an obvious command line flag to tell NPM to look elsewhere for project-level config (which makes sense), so the only way I think you could make this work would be:

1. Hydrate checks for `.npmrc` in the function folder, if exists, do nothing.
2. Else, check for `.npmrc` in the "arc root" (I'm not sure this exists as a concept now function code can live anywhere, so it's not as simple as `../..`?), if exists, copy/link to the current path, do an install, remove the link/file again.

What problems might this introduce? The only (big!) one I can think of is that a project may have had an `.npmrc` at project root for a while that has only impacted sandbox's operation, but not deployment. With this change it will impact both. Make it a deployment configurable option, off by default? Get sandbox/hydate/most appropriate script to show a warning about what might happen, confirm on first time?

**Comment by ryanblock** [comment]
Right! Re 2, that approach is what I was obliquely referring to when I said above `I don't currently have any elegant ideas of how to handle this, especially considering the side effects that may occur if we start doing things with the .nmprc file within the project.` A goal I've maintained around Lambda treeshaking is to only mutate a function's `node_modules` dir; I'm not even comfortable putting a temporary generated `package.json` in there (but I don't have a way around it).

`.npmrc` files also often have auth tokens, which would automatically opt them out of being checked in; and if they were gitignored in the root, but then automatically copied into an individual function, and a deploy failed unexpectedly, an `.npmrc` file with creds could be checked into the repo.

This is kind of why I think it's best off as a userland problem: no matter what, future developers will need some kind of secret to access private npm registries, which means they're already going to have an onboarding step associated. The more I think through this, the more certain I'm becoming that private NPM registry config shouldn't be an Arc responsibility.

**Comment by andybee** [comment]
I accept this is better being a user land issue, but for reference the way I set this up is that the committed `.npmrc` contains only the reference that `x` package namespace exists on `y` repo URL, not the credentials themselves.

This means when a new dev first runs `npm install` you get a helpful bump of that you're not authorised to pull that package and therefore have a clue to go seek something out. When you authorise using npm it drops the credentials in to `~` so it's off the project.

I prefer this approach to missing it off entirely and dealing with a "package not found" error when running `npm install`, I'm happy to at least provide the true location to anyone who has the codebase and give them a clue they probably need to authenticate. As much as I'm sure everyone reads the dev environment README I painstakingly put together... 😉

**Comment by filmaj** [comment]
Closing this as I think the end result here is "userland concern." I wonder if symlinks can also helps here, e.g. symlinking `src/shared/.npmrc` to project root `.npmrc`.
